core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -328,8 +328,9 @@ class ControllerContext {
   }
 
   def queueTopicDeletion(topics: Set[String]): Unit = {
+    val newlyDeletedTopics = topics.diff(topicsToBeDeleted)
```

Comment:
1. Could we add a comment here to mention why we need `diff` here?
2. I don't think the parameter name `topics` is clear enough. Could we update it? Ex: topics -> `newTopicsToBeDeleted` or `topicToBeAddedIntoDeletionList` or something else?

Reply:
One question: After this change, `topic-A` in your example in JIRA will be only entered into `cleanPreferredReplicaImbalanceMetric` once, right? So, what if when the first time entering `cleanPreferredReplicaImbalanceMetric`, it haven't run topic-A's deletion procedure to mark topic-A's all partitions as Offline (Leader = -1) here: https://github.com/apache/kafka/blob/3.2.0/core/src/main/scala/kafka/controller/ReplicaStateMachine.scala#L368 ? I mean, since we only decrement `preferredReplicaImbalanceCount` if topic-A has NoLeader and `!hasPreferredLeader(replicaAssignment, leadershipInfo)` evaluates to true, right? So, what if the topic-A haven't reached this condition and we entering `cleanPreferredReplicaImbalanceMetric`? Will that cause some partitions count should be decremented but they are not? Or I should say, will that case happen?

Reply:
(I'm not 100% sure if I get your concern. If my answer doesn't make sense, please tell me)

In the first place, the purpose of `cleanPreferredReplicaImbalanceMetric` is to "unmark" partitions which has imbalanced leader.

So, below is the expected behavior (let's say topic-A has single partition):
- Initiate topic-A deletion
- If topic-A-0 has imbalanced leader (e.g. lost leadership once, replica-reassignment, became offline etc) here,  `preferredReplicaImbalanceCount` contains the count for it here, so we have to decrement it because the partition is to be deleted
  * `!hasPreferredLeader` evaluates to true so decremented as expected
- Otherwise, we don't have to decrement the value because `preferredReplicaImbalanceCount` doesn't contain the imbalance count for `topic-A-0`.

"A partition should be decremented from preferredReplicaImbalanceCount on topic-deletion only if the partition is actually imbalanced at the topic-deletion timing" is the invariant that topic-deletion procedure must satisfy.

So, essentially, the problem is that topic-A-0 could be marked as "imbalanced" wrongly when it becomes offline during deletion-procedure.

> So, what if the topic-A haven't reached this condition and we entering cleanPreferredReplicaImbalanceMetric? Will that cause some partitions count should be decremented but they are not?

"topic-A haven't reached this condition" means the partition wasn't imbalanced actually, so we do not have to decrement the count.

Reply:
@ocadaruma , thanks for the explanation. So, do you mean:
1. Before the PR, topic-A partition 0 is in preferred Leader, so after 1st time entering `cleanPreferredReplicaImbalanceMetric`, it won't do anything, which is expected, because ithe `preferredReplicaImbalanceCount` is not incremented before. But after 1st `topic-A` deletion, the leader will be changed to NoLeader (or other replica), and now, if we have 2nd chance entering  `cleanPreferredReplicaImbalanceMetric`, it'll decrement it, which is unexpected. Is my understanding correct?

Reply:
@showuon Correct.

Reply:
I understand now. Thanks.